[windows][usm] load apm tags along with usm service tags

derekwbrown opened the pull request:
### What does this PR do?

On IIS (only at this time), load the APM config from either `datadog.json` or `web.config`, and add
those tags to the `dynamictags` list that's sent to the back end

### Motivation

be able to correlate USM services with APM services

### Additional Notes

### Possible Drawbacks / Trade-offs


### Describe how to test/QA your changes
Install agent.
Install IIS.
Create one or more IIS sites in the IIS configuration.

Place a `datadog.json` or `web.config` file in the root of the site directory, with any combination of env/version/service tags.
The file formats are documented
https://datadoghq.atlassian.net/wiki/spaces/WKIT/pages/4012835378/USM+APM+Compatibility#.NET-Tracer-configuration-file-formats

Make a connection to the site.

YOu can check that the appropriate tag(s) have been added by hitting the HTTP information endpoint:
((iwr -UseBasicParsing -DisableKeepAlive http://localhost:3333/network_tracer/debug/http_monitoring).content | ConvertFrom-Json)

YOu should see output such as:

```

Client      : @{IP=::1; Port=80}
Server      : @{IP=::1; Port=62700}
DNS         :
Path        : /
Method      : GET
ByStatus    : @{200=}
StaticTags  : 0
DynamicTags : {http.iis.app_pool:DefaultAppPool, http.iis.site:1, http.iis.sitename:Default Web Site}
```

with the addition of `service`, `env`, and or `version` tags that match what was configured in the file that you applied.

derekwbrown committed:
Move `iisconfig` to its own package under winutil.  We're going to be
adding a bunch of other types which will just pollute the `winutil` package

derekwbrown committed:
Add parsing of the files.

Add tests to make sure we process things correctly

derekwbrown committed:
Add apm tags to dynamic tags of http connection

derekwbrown committed:
fix lazy locking during config initialization

derekwbrown committed:
add comments for linting

iglendd commented on pkg/network/protocols/http/etw_http_service.go:
```diff
@@ -1427,7 +1433,7 @@ func (hei *EtwInterface) OnStart() {
 	httpServiceSubscribed = true
 	var err error
-	iisConfig, err = winutil.NewDynamicIISConfig()
+	iisConfig, err = iisconfig.NewDynamicIISConfig()
```
Yeah, from the previous comment iisConfig may not be fully initialized if at all and in theory you may crash on a first event.

iglendd replied:
that is my theory of course :)

derekwbrown replied:
from previous comment, iisconfig has its own locking to ensure it's consistent.

derekwbrown replied:
ah, ok.  Now I get what you're saying.
And yes you're right (and worse, if the iisconfig init returns an error, we'll still dereference the nil pointer as you point out).  Fixing.

iglendd commented on pkg/util/winutil/iisconfig/tagtree.go:
```diff
@@ -0,0 +1,157 @@
+
+	"golang.org/x/sys/windows/registry"
+)
+
```
Can you add a comment explaining how pathTree(Entry) works. It is not entirely clear

iglendd commented on pkg/util/winutil/iisconfig/tagtree.go:
```diff
@@ -0,0 +1,157 @@
+	return append(s, b)
+}
+
+func findInPathTree(pathtrees map[uint32]*pathTreeEntry, id uint32, urlpath string) (APMTags, APMTags) {
```
should id be called siteid since this terminology is used in other places?

iglendd commented on pkg/util/winutil/iisconfig/tagtree.go:
```diff
@@ -0,0 +1,157 @@
+
+	currNode := pathtrees[id]
+
+	for _, part := range pathparts {
+		if _, ok := currNode.nodes[part]; !ok {
+			currNode.nodes[part] = &pathTreeEntry{
+				nodes: make(map[string]*pathTreeEntry),
+			}
+		}
+		currNode = currNode.nodes[part]
+	}
+	currNode.ddjson = ddjson
+	currNode.appconfig = appconfig
+	return
```
I wonder why would you need to split on pars if whole path can be used as a key?

derekwbrown replied:
What I've done is made a search tree.
As is demonstrated in the test
/path/to/something is inside of /path.  It's inside of /path/to  but it's not in /path/to/else

if you just search on entire path, you'll get false matches.

iglendd commented on pkg/util/winutil/iisconfig/tagtree.go:
```diff
@@ -0,0 +1,157 @@
+
+}
+func (iiscfg *DynamicIISConfig) GetAPMTags(siteID uint32, urlpath string) (APMTags, APMTags) {
+	iiscfg.mux.Lock()
+	defer iiscfg.mux.Unlock()
```
In current code iiscfg iself may not be initialized at early stages

derekwbrown replied:
that locking is handled in iisconfig; added a test to demonstrate.

iglendd reviewed (changes requested):
Some clarification is requested

derekwbrown committed:
Change APM tags to the previously established tags, and only send
highest precedence rather than sending all that we found.

derekwbrown commented on pkg/network/protocols/http/etw_http_service.go:
```diff
@@ -742,6 +744,8 @@ func httpCallbackOnHTTPRequestTraceTaskParse(eventInfo *etw.DDEventRecord) {
 		if len(urlParsed.Path) == 0 {
 			urlParsed.Path = "/"
 		}
+		httpConnLink.urlPath = urlParsed.Path
```
the RequestFragment is constructed by us so that it matches what the shared code is expecting later in the processing.
it matches what the linux/ebpf side uses.

iglendd replied:
right

derekwbrown commented on pkg/network/protocols/http/etw_http_service.go:
```diff
@@ -832,6 +836,8 @@ func httpCallbackOnHTTPRequestTraceTaskDeliver(eventInfo *etw.DDEventRecord) {
 	httpConnLink.http.SiteID = userData.GetUint32(16)
 	httpConnLink.http.SiteName = iisConfig.GetSiteNameFromID(httpConnLink.http.SiteID)
 
+	httpConnLink.http.TagsFromJson, httpConnLink.http.TagsFromConfig = iisConfig.GetAPMTags(httpConnLink.http.SiteID, httpConnLink.urlPath)
```
iisconfig has its own locking to ensure that it's consistent.

derekwbrown committed:
fix incorrect conditional

derekwbrown commented on pkg/network/protocols/http/model_windows.go:
```diff
@@ -107,11 +107,30 @@ func (tx *WinHttpTransaction) StaticTags() uint64 {
+		if (len(tx.TagsFromJson.DDVersion)) > 0 {
+			tags = append(tags, fmt.Sprintf("http.iis.cfg.ddversion:%v", tx.TagsFromJson.DDVersion))
+		}
+		return tags
```
oh, good catch.
I had updated the block because of an unrelated comment, but that logic error still existed.

derekwbrown commented on pkg/util/winutil/iisconfig/apmtags.go:
```diff
@@ -0,0 +1,102 @@
+		}
+	}
+	if len(chaseDatadogJson) > 0 {
+		ddjson, err := ReadDatadogJson(chaseDatadogJson)
```
the problem is that this code will potentially be executed on every process start.  Logging (without some sort of throttling) would just flood the log file.

derekwbrown committed:
add test to demonstrate race safety

derekwbrown committed:
go mod tidy; fix incorrect slice append

iglendd commented on pkg/util/winutil/iisconfig/tagtree.go:
```diff
@@ -0,0 +1,157 @@
+func findInPathTree(pathtrees map[uint32]*pathTreeEntry, id uint32, urlpath string) (APMTags, APMTags) {
+	// urlpath will come in as something like
+	// /path/to/app
+	// need to build the tree all the way down
```
wrong comment (at least last line comment)

iglendd commented on pkg/util/winutil/iisconfig/tagtree.go:
```diff
@@ -0,0 +1,157 @@
+	currNode := pathtrees[id]
+
+	for _, part := range pathparts {
+		if _, ok := currNode.nodes[part]; !ok {
+			currNode.nodes[part] = &pathTreeEntry{
+				nodes: make(map[string]*pathTreeEntry),
+			}
+		}
+		currNode = currNode.nodes[part]
```
```suggestion
		if _, ok := currNode.nodes[part]; !ok {
			currNode = &pathTreeEntry{ nodes: make(map[string]*pathTreeEntry),}
			currNode.nodes[part] = cureNode
		} else {
		    currNode = currNode.nodes[part]
		}
```

derekwbrown committed:
fix linter; review feedback

derekwbrown committed:
fix previous replace errors; fix pre-existing race condition on IIS configuration object

derekwbrown committed:
change variable name

iglendd reviewed (approved):
Looks good. Left a few optional suggestions

derekwbrown committed:
remove incorrect comment

alexn-dd commented on pkg/network/protocols/http/etw_http_service.go:
```diff
@@ -977,7 +986,10 @@ func httpCallbackOnHTTPRequestTraceTaskSrvdFrmCache(eventInfo *etw.DDEventRecord
-		httpConnLink.http.SiteName = iisConfig.GetSiteNameFromID(cacheEntry.http.SiteID)
+		cfg := iisConfig.Load()
+		if cfg != nil {
+			httpConnLink.http.SiteName = cfg.GetSiteNameFromID(cacheEntry.http.SiteID)
```
So this means web.config overrides datadog.json, correct?

alexn-dd commented on pkg/network/protocols/http/etw_http_service.go:
```diff
@@ -977,7 +986,10 @@ func httpCallbackOnHTTPRequestTraceTaskSrvdFrmCache(eventInfo *etw.DDEventRecord
-		httpConnLink.http.SiteName = iisConfig.GetSiteNameFromID(cacheEntry.http.SiteID)
+		cfg := iisConfig.Load()
+		if cfg != nil {
+			httpConnLink.http.SiteName = cfg.GetSiteNameFromID(cacheEntry.http.SiteID)
```
Curious, why does the callback need to load the IIS config when fetching values from a cache?

derekwbrown replied:
Good question, which I'm not totally sure.
Note this behavior is preexisting; the change here is to prevent dereferencing a `nil` pointer.

iglendd replied:
I do not think it was me. I think, if remember my 2 years old implementation that pool name came with ETW payload as well as SiteID but not SiteName (just above `httpConnLink.http.SiteID = userData.GetUint32(16)`), I think. I think it was you Derek who added correlating SiteID to the Site name a while ago by getting it from a map loaded from config

iglendd replied:
Ah, I think Alex, got fooled by load name. It is not actually load in the sense initialization per se but atomic copying from a variable. Are we on the same page?

derekwbrown replied:
resolved (I think) in slack

alexn-dd commented on pkg/network/protocols/http/model_windows.go:
```diff
@@ -106,14 +106,37 @@ func (tx *WinHttpTransaction) StaticTags() uint64 {
+		tags = append(tags, fmt.Sprintf("http.iis.sitename:%v", tx.SiteName))
+	}
+
+	// tag precedence is web.config -> datadog.json
```
I hope this detail doesn't get lost.

derekwbrown replied:
The internal documentation has the full description of the precedence.

alexn-dd commented on pkg/network/protocols/http/model_windows.go:
```diff
@@ -106,14 +106,37 @@ func (tx *WinHttpTransaction) StaticTags() uint64 {
+	} else if (len(tx.TagsFromJson.DDVersion)) > 0 {
+		tags = append(tags, fmt.Sprintf("version:%v", tx.TagsFromJson.DDVersion))
+	}
+	if len(tags) == 0 {
```
This is dead code this tags is guaranteed to at least have: fmt.Sprintf("http.iis.site:%v", tx.SiteID)

derekwbrown replied:
good catch.  I accidentally changed the behavior above when fixing another subtle request.
Code is no longer dead.

derekwbrown committed:
add log on error case

alexn-dd commented on pkg/util/winutil/iisconfig/apmtags.go:
```diff
@@ -0,0 +1,103 @@
+}
+type iisAppSettings struct {
+	XMLName xml.Name        `xml:"appSettings"`
+	Adds    []iisAppSetting `xml:"add"`
```
Curious, can this be a map if search is expected?
If there is no search on this, a list is fine.

derekwbrown replied:
I'm not following.  it's a list because that's what the xml is providing.  the `xml` tag allows the xml decoder to marshal the xml straight into the structure, so the structure needs to follow the xml.

alexn-dd commented on pkg/util/winutil/iisconfig/apmtags.go:
```diff
@@ -0,0 +1,103 @@
+	if err != nil {
+		return apmtags, err
+	}
+	err = xml.Unmarshal(f, &newcfg)
```
This is a bit funny.  Whoever wrote Unmarshal must be a C/C++ person.

alexn-dd commented on pkg/util/winutil/iisconfig/apmtags.go:
```diff
@@ -0,0 +1,103 @@
+			chasedatadogJSON = setting.Value
+		}
+	}
+	if len(chasedatadogJSON) > 0 {
```
I suggest adding a comment here, that this is getting datadog.json embedded within the IIS config.

alexn-dd commented on pkg/util/winutil/iisconfig/apmtags.go:
```diff
@@ -0,0 +1,103 @@
+	}
+	if len(chasedatadogJSON) > 0 {
+		ddjson, err := ReadDatadogJSON(chasedatadogJSON)
+		if err == nil {
```
Hmm, this seems to contradict the previous preference of web.config > datadog.json.

derekwbrown replied:
it  doesn't.  And it's spelled out in the internal doc.
there's a precedence order.  The web config can contain an entry pointing to the datadog.json, in which case we chase the "pointer".

alexn-dd commented on pkg/util/winutil/iisconfig/apmtags_test.go:
```diff
@@ -0,0 +1,117 @@
+		assert.Equal(t, "to", sp[1])
+		assert.Equal(t, "app", sp[2])
+	})
+
```
I also recommend testing multiple separators: `/foo///` and empty ``

derekwbrown replied:
These are URL paths. are those valid URL paths?
(btw, the fact that it's URL paths is why I had to write my own `splitPaths`.
URL path separators are expected to be `/`; the go path split since we're running on windows is expecting the `\` separators. .  The implementation of `splitPaths` also calls `clean` to handle the rest.

alexn-dd commented on pkg/util/winutil/iisconfig/apmtags_test.go:
```diff
@@ -0,0 +1,117 @@
+	os.Setenv("TESTROOTDIR", testroot)
+	defer os.Unsetenv("TESTROOTDIR")
+
+	//apppath := filepath.Join(path, "testdata", "iisconfig.xml")
```
Dead code

derekwbrown replied:
done.

alexn-dd commented on pkg/util/winutil/iisconfig/apmtags_test.go:
```diff
@@ -0,0 +1,117 @@
+		assert.Equal(t, "app1", tags.DDService)
+	})
+	t.Run("test secondary site actual app3", func(t *testing.T) {
+		tags, _ := iisCfg.GetAPMTags(3, "/siteapp2/siteapp3")
```
Does it matter if the path is to a file?

derekwbrown replied:
shouldn't, but I added that case.

alexn-dd commented on pkg/util/winutil/iisconfig/iisconfig.go:
```diff
@@ -34,6 +34,7 @@ type DynamicIISConfig struct {
 	stopChannel  chan bool
 	xmlcfg       *iisConfiguration
 	siteIDToName map[uint32]string
+	pathtrees    map[uint32]*pathTreeEntry
```
Nit: pathTrees

derekwbrown replied:
fixed

derekwbrown committed:
Fix inadvertent change in behavior.  Only add the site id if we were able
to decode the app pool or the site name.

derekwbrown committed:
Review feedback

Fixed a logic change found in review.
Added an additional test.
and other assorted changes.

alexn-dd commented on pkg/util/winutil/iisconfig/apmtags.go:
```diff
@@ -0,0 +1,103 @@
+	if err != nil {
+		return apmtags, err
+	}
+	apmtags.DDService = datadogJSON["DD_SERVICE"]
```
Since these fields are user editable, should they be sanitized?
I don't know what guarantees Go provides .

derekwbrown replied:
sanitized in what way?  This should be configured by the customer to be the service value they want to send.
(our tags will be post-processed later to remove special chars, etc.)

alexn-dd replied:
OK, so if there is no processing of the values (e.g. copy-pasted strings), should be fine.

alexn-dd commented on pkg/util/winutil/iisconfig/apmtags.go:
```diff
@@ -0,0 +1,103 @@
+	for _, setting := range newcfg.AppSettings.Adds {
+		switch setting.Key {
+		case "DD_SERVICE":
+			apmtags.DDService = setting.Value
```
Since web.config is an external (modifiable) data source, should these be sanitized?

derekwbrown replied:
this particular value is to be configured with exactly what the customer wants to send to us.

pgimalac reviewed (approved):
LGTM for ASC files

derekwbrown committed:
linter

derekwbrown commented:
/merge

derekwbrown commented:
/merge
